[DTensor] Error on illegal view op during sharding prop

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149764
* #152045

Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

add 'strict_view' kwarg to the helper methods that implement
view/reshape op shard prop rules, so it can be decided op-by-op whether
to raise these new errors
enabled errors just for the 'view' op in this PR
added two specific checks/errors that can occur during view ops.

Details:

- View ops are never allowed to flatten a dimension that is unevenly
  sharded, since that would likely change the size/content of the
  local_tensor and require redistribute
- View ops are also never allowed to flatten two dims if the rightmost
  dim is a Shard() placment, becuase it would cause contiguity errors
  without redistribution

Notes:

- Disables support for several ops in test_dtensor_ops.py test, which
  decompose to an illegal view that only works by performing a
  redistribution: cartesian_prod, flatten, ravel, reshape, reshape_as, view, view_as, take_along_dim, kron

Follow Ups:
- triage other view-like ops (besides aten::view) for using strict_view
- look for other gaps where view-like ops could still perform
  redistribution (ban them all, and document this)

Fixes #143372

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k

wconstab committed:
[DTensor] Error on illegal view op sharding prop

Following the definition of legal sharding in the previous PR, this
enforces  that view ops raise a clear error message if an invalid
uneven sharding is encountered during view op sharding propagation.

The mechanism is to compute the current local shapes based on the
existing input sharding, and  the new local shapes based on the proposed
output sharding and new global shape.  The new global shape is
pre-existing, but the computation of new local output shape is new and
requires additional infrastructure, since normally we'd just run the
aten operator to get the new shape, and in this case we can't even do
that since we'd risk performing an impossible view (e.g. tensor of shape
(512,) reshaped locally to (384,) in the particular example being
addressed here.

Fixes #143372

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -442,6 +452,21 @@ def dim_reduction(
     )
 
 
+def compute_shape(input_shape: Shape, output_rules: DimMap) -> Shape:
```
what do folks think about this new util- is it worth adding, and OK with my approach of adding 'concrete_shape' fields to the different DimSpec impls?  I would complete the others too but did not do it yet.

wanchaol reviewed (changes requested):
Hmmm I don't quite understand this PR, why we need to have those global/local shape calculations/checks in order to error out on illegal views (and only error out in certain cases)?

IMO the changes to error out should be very simple:
1. add the `strict_view` option as this PR did.
2. inside `reshape_strategy`, just check if `strict_view` and `input_src_spec != input_target_spec`, then error out.

We should not rely on any shape checks to determine whether we want to error out, it should be just `view` ops should not do any redistribute

wconstab commented:
> inside reshape_strategy, just check if strict_view and input_src_spec != input_target_spec, then error out.

In the concrete example i'm trying to fix, the input_src_spec == input_target_spec == (Shard(0)).

Both spec are exactly the same:
```
DTensorSpec(mesh=DeviceMesh('cuda', [0, 1, 2, 3], mesh_dim_names=('tp',)), placements=(Shard(dim=0),), tensor_meta=TensorMeta(shape=torch.Size(
[6, 256]), stride=(256, 1), dtype=torch.float32))
```

The problem is, before the view, the local tensors would be 
Rank0-2: (2,256)
Rank3: (0, 256)
After the view, the local tensors 'should' be 
Rank0-3: (384,)
But there is nothing keeping track of this inside the spec, and in reality the local data is more like 
Rank0-2: (512,)
Rank3: (0,)

I could not think of a way to raise the error inside the view sharding prop rule, unless I compute the concrete local shapes.  Do you have a suggestion of another way to do this?

wanchaol commented:
> 
> The problem is, before the view, the local tensors would be Rank0-2: (2,256) Rank3: (0, 256) After the view, the local tensors 'should' be Rank0-3: (384,) But there is nothing keeping track of this inside the spec, and in reality the local data is more like Rank0-2: (512,) Rank3: (0,)
> 
> I could not think of a way to raise the error inside the view sharding prop rule, unless I compute the concrete local shapes. Do you have a suggestion of another way to do this?

Oh I see, I think for this case we could check in the Flatten operation around here https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_view_ops.py#L515. Basically if we want to make sure the first dim size of the flattened dimensions is evenly divisible to the mesh size, then it's "shardable". If not then we should error out with illegal views.

wconstab commented:
> Oh I see, I think for this case we could check in the Flatten operation around here https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_view_ops.py#L515. Basically if we want to make sure the first dim size of the flattened dimensions is evenly divisible to the mesh size, then it's "shardable". If not then we should error out with illegal views. The Split operation have some similar checks

This makes sense. But how can I compute the size of the flattened dimension to make sure it is divisible by the mesh size?  IIUC the reason this can be computed in the 'split' case is becuase when you're doing a split you have to pass the explicit new split sizes as part of the command, but for flatten you can use `-1`.  I did put logic into DimSpec for computing 'concrete_shape' to solve this problem,  but it sounds like that's not the way you had in mind?

wanchaol commented:
> This makes sense. But how can I compute the size of the flattened dimension to make sure it is divisible by the mesh size? IIUC the reason this can be computed in the 'split' case is because when you're doing a split you have to pass the explicit new split sizes as part of the command, but for flatten you can use `-1`. I did put logic into DimSpec for computing 'concrete_shape' to solve this problem, but it sounds like that's not the way you had in mind?

I think The DimSpec transformation thing is already able to get rid of `-1` and give concrete dims? i.e. whenever you are in the `Flatten` cmd, the `cmd.input_dims` should all be `InputDim` that have concrete input dimensions already. For the shape/size of flattened dimension, because view is a complicated transformation, computing the full shape would not be always correct. I think what we should do is to add the `group_shape` to `Flatten` similar to `Split` here https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_ops/_view_ops.py#L368, and then when dealing with Flatten we know: 1. the from_shape generated in the process of decomposing views 2. the first flattened dim from the `Flatten` DimSpec. Then we can throw error correspondingly.

wconstab commented:
Thinking about this more, I see a few places I could do this error checking, but they have tradeoffs.

1) `view_groups` function.  This is where flatten is constructed.  In this function, I already know the old global shape (6,256) and new global shape (starts as -1, gets inferred to 6*256), but I do not know the mesh dim shape and sharding spec.  For example, I do not know that the tensor spec says 'shard(0)' or the size of the first mesh dim is '4', which is critical for determining whether the sharding will be uneven.

2) `register_op_strategy_map:reshape_strategy`:  This fn knows the op type (useful for asserting on specific op type, like 'strict_view' field.  It also knows the mesh shape, so it solves the problem in (1). But it does not know the concrete shapes at all.  It has access to the 'rules' which are DimSpec.

3) propagate_shape_and_sharding: This fn knows the mesh_shape, and also has access to the rules just like (2).  It also has logic specific to each rule type (e.g. Flatten) so there is a logical place to handle the case of flattening.  A flag like 'strict_view' could also be passed into this function if needed.

For (2) and (3), i think the only missing piece is adding some shape information to the Flatten object.  IIUC that's what you are proposing here:
> I think what we should do is to add the group_shape to Flatten similar to Split

I'll update the PR to try this and lmk if you have a preference between 1,2,3.  Thanks for the suggestions!

wanchaol commented:
> 
> Edit: Its easy to add the 'group shape' to the Flatten. However, i'm not sure its good to raise the error locally when looking at the 'flatten' op. Would it be possible to have a complex view that first creates an 'incompatible' flatten, but then splits and the final result is 'compatible'? If so, then I think the safest place to error is at (2) based on the original/final shape or original/final nelem.

IMO the right way to do this is to error when seeing a local Flatten DimSpec (i.e. add a group_shape to Flatten), The whole DimSpec logic is there to figure out a viable transformation steps for view like operations, this means that the dims need to actually go through those transformations conceptually in order to become the "output shape". So I think for the case of a complex view that have a "incompatible" flatten in the middle, and if this flatten need to error, then the whole transformation would not be compatible and should error too. However on the other side, if we handle it directly with the old concrete/global shape might not cover some complex view transformations.

So I think handling Flatten inside 3. `propagate_shape_and_sharding` should give the correct behaviors.

wconstab commented:
> IMO the right way to do this is to error when seeing a local Flatten DimSpec (i.e. add a group_shape to Flatten), 

@wanchaol after thinking more about this, I think that using 'shape' (or group_shape) might not be the right approach.  Or at least, I am not sure what criteria to apply to 'shape' to determine if flattening is illegal.  Initially, I planned to assert that the new local shape matches the existing local shard shape, but i now believe this is not correct.  Example:  I could flatten and split some sharded dims, which is a violation, but then i could create a local shape that looks the same as the original shape and not catch this error by shape-checking.

Instead, I think the best approach is to propagate another piece of information, about input dim sharding.  The rules would be recursive rules.  

Valid to flatten if: 
        - all replica dims
        - a shard on the leftmost side, and replicated dims to the right

        Invalid, if
        - two shard dims (except if they were created by a split on a shard dim)
            - i showed a couple examples where this is definitely invalid, but i'm not sure if its gauranteed to be invalid?
        - one or more replicas on the left followed by a shard on the right
And the way to implement these rules could be recursive propagation on the different helper classes
      - Split: Shard -> (Shard, Shard),  Replica -> (Replica, Replica)
      - Flatten: (Shard, Replicate...) -> Shard,  (Replicate, ...) -> Replicate

Does this idea make sense? Might be easier to discuss over chat or VC if not.

wconstab committed:
Update on "[WIP][DTensor] Error on illegal view op sharding prop"


Following the definition of legal sharding in the previous PR, this
enforces  that view ops raise a clear error message if an invalid
uneven sharding is encountered during view op sharding propagation.

The mechanism is to compute the current local shapes based on the
existing input sharding, and  the new local shapes based on the proposed
output sharding and new global shape.  The new global shape is
pre-existing, but the computation of new local output shape is new and
requires additional infrastructure, since normally we'd just run the
aten operator to get the new shape, and in this case we can't even do
that since we'd risk performing an impossible view (e.g. tensor of shape
(512,) reshaped locally to (384,) in the particular example being
addressed here.

Fixes #143372

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Details:

Fixes #143372

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.  

The motivation is twofold:
1. to provide clearer errors in currently failing cases
2. to clarify that _becuase_ view ops are expected to be lightweight ops, it would be bad to implicitly perform communication that would be needed to support these failing cases, and thus the right thing is to error clearly.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

* add 'strict_view' kwarg to the helper methods that implement view/reshape op shard prop rules, so it can be decided op-by-op whether to raise these new errors
* enabled errors just for the 'view' op in this PR
* added two specific checks/errors that can occur during view ops.  

Details about new errors added for 'strict_view' ops:
(1) a Flatten is only allowed if only the left-most input dim is sharded or all input dims are replicated
(2) when Flatten is applied on a sharded left-most input dim that is unevenly sharded, the Flatten is also illegal since it would likely change the unevenness and require resharding.


Follow ups
* we should decide what other view-like ops behavior should be (this PR only addressed the 'view' op

Fixes #143372

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.  

The motivation is twofold:
1. to provide clearer errors in currently failing cases
2. to clarify that _becuase_ view ops are expected to be lightweight ops, it would be bad to implicitly perform communication that would be needed to support these failing cases, and thus the right thing is to error clearly.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

* add 'strict_view' kwarg to the helper methods that implement view/reshape op shard prop rules, so it can be decided op-by-op whether to raise these new errors
* enabled errors just for the 'view' op in this PR
* added two specific checks/errors that can occur during view ops.  

Details about new errors added for 'strict_view' ops:
(1) a Flatten is only allowed if only the left-most input dim is sharded or all input dims are replicated
(2) when Flatten is applied on a sharded left-most input dim that is unevenly sharded, the Flatten is also illegal since it would likely change the unevenness and require resharding.


Follow ups
* we should decide what other view-like ops behavior should be (this PR only addressed the 'view' op

Fixes #143372

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.  

The motivation is twofold:
1. to provide clearer errors in currently failing cases
2. to clarify that _becuase_ view ops are expected to be lightweight ops, it would be bad to implicitly perform communication that would be needed to support these failing cases, and thus the right thing is to error clearly.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

* add 'strict_view' kwarg to the helper methods that implement view/reshape op shard prop rules, so it can be decided op-by-op whether to raise these new errors
* enabled errors just for the 'view' op in this PR
* added two specific checks/errors that can occur during view ops.  

Details about new errors added for 'strict_view' ops:
(1) a Flatten is only allowed if only the left-most input dim is sharded or all input dims are replicated
(2) when Flatten is applied on a sharded left-most input dim that is unevenly sharded, the Flatten is also illegal since it would likely change the unevenness and require resharding.


Follow ups
* we should decide what other view-like ops behavior should be (this PR only addressed the 'view' op

Fixes #143372

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wanchaol reviewed (commented):
The changes looks good and almost ready to me! Have some additional comments inlined

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -502,17 +502,59 @@ def collect_used_inputs(cmd: DimSpec) -> None:
+    for dim in range(len(global_input_shape)):
         shardable_dims[dim] = [dim in seen_input_dims] * mesh_ndim
 
+    def is_sharded(input_dim: InputDim) -> bool:
+        for placement in input_src_placements:
+            if isinstance(placement, Shard) and placement.dim == input_dim.input_dim:
+                return True
+        return False
+
+    def get_shard_mesh_dim_and_placement(
+        input_dim: InputDim,
+    ) -> tuple[int, Shard]:
+        # if input_dim is sharded, return the mesh_dim and shard placement
+        for i, placement in enumerate(input_src_placements):
+            if isinstance(placement, Shard) and placement.dim == input_dim.input_dim:
+                return i, placement
+        raise RuntimeError("input_dim is not sharded")
+
```
haha. i did that originally, but it was kind of awkward. i can try to refactor a bit more.

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -608,10 +661,11 @@ def reshape_strategy(op_schema: OpSchema) -> StrategyType:
-            ]
+            redistribute_costs: list[list[float]] = []
 
+            # TODO(whc) should attach propagated tensor_meta.
```
i see. i think @XilunWu and I were reading the code together a few weeks ago and this looked like a gap.  We were already on the looking out for this because @fmassa had complained that some sharding prop were missing tensor_meta.  If in this case there is a later step where we attach it, then thats great. maybe i will just update with a comment here explaining that.  Do you know where that code is?

wanchaol replied:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/_sharding_prop.py#L279

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -608,10 +661,11 @@ def reshape_strategy(op_schema: OpSchema) -> StrategyType:
-            redistribute_costs = [
-                generate_redistribute_costs(input_strategy, input_tgt_spec)
-            ]
+            redistribute_costs: list[list[float]] = []
```
oh no, this looks like a complete accident. I think i was probably just trying to add the type information and then the list content got accidentally deleted.  I'm surprised it doesn't break something 🤔

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -502,17 +502,59 @@ def collect_used_inputs(cmd: DimSpec) -> None:
+                        shard_mesh_dim, shard_placement = (
+                            get_shard_mesh_dim_and_placement(dim)
+                        )
+                        tensor_dim_size = global_input_shape[shard_placement.dim]
```
hmm. the way I thought about it, I thought we _need_ to check the global shape.  That is the way to determine how the tensor was originally sharded before the view.  It determines whether the current local_tensor is evenly or unevenly sharded, for example.

we can also check the 'group_shape', but it is possible the group_shape can be evenly shardable while the original global shape is not, and we need to catch this case.

Example:

tensor shape (2, 2)
mesh shape (4)
if we shard the tensor [Shard(0)] and then we view (-1), the global shape would show us unevenness, but the flattened shape would show even sharding.

wanchaol replied:
I see. but for this case the group_shape would be (2, 2), and it would just error out before the flatten is done right, this is also before the actual view happened?

wconstab replied:
we should probably align on what the definition is for "Flatten.group_shape".

my assumption is, it is the 'new shape of the dimension' after the flatten.  So if the input shape of the tensor is (2, 2) and we flatten dims 0,1, I think the new group_shape would be (4,).  Is that also how you imagined it, or what?

> the group_shape would be (2, 2)

in your comment it sounds like the opposite.  group_shape is "the old shape before the flatten"?  I think this way is more useful, becuase at least (2, 2) can be used to detect the uneven sharding condition.  But it also feels inconsistent with Split.group_shape, which IIUC is the 'new shape after split' rather than the original dim size before split.

wanchaol replied:
Oh yeah, the `group_shape` is the old shape before the flatten, Split.group_shape is the new shape after split as it need to record the new shape to determine sharding after Split. Because Flatten is a reverse operation compare to Split, it should also record the same before flatten. Maybe we can make it more clear with a different name if that confuses you

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.  

The motivation is twofold:
1. to provide clearer errors in currently failing cases
2. to clarify that _becuase_ view ops are expected to be lightweight ops, it would be bad to implicitly perform communication that would be needed to support these failing cases, and thus the right thing is to error clearly.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

* add 'strict_view' kwarg to the helper methods that implement view/reshape op shard prop rules, so it can be decided op-by-op whether to raise these new errors
* enabled errors just for the 'view' op in this PR
* added two specific checks/errors that can occur during view ops.  

Details about new errors added for 'strict_view' ops:
(1) a Flatten is only allowed if only the left-most input dim is sharded or all input dims are replicated
(2) when Flatten is applied on a sharded left-most input dim that is unevenly sharded, the Flatten is also illegal since it would likely change the unevenness and require resharding.


Follow ups
* we should decide what other view-like ops behavior should be (this PR only addressed the 'view' op

Fixes #143372

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wanchaol reviewed (approved):
This looks good to go, we can address the group_shape comment as a follow up if needed. Please remove the tensor meta comment and maybe just with a comment to explain

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.  

The motivation is twofold:
1. to provide clearer errors in currently failing cases
2. to clarify that _becuase_ view ops are expected to be lightweight ops, it would be bad to implicitly perform communication that would be needed to support these failing cases, and thus the right thing is to error clearly.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

* add 'strict_view' kwarg to the helper methods that implement view/reshape op shard prop rules, so it can be decided op-by-op whether to raise these new errors
* enabled errors just for the 'view' op in this PR
* added two specific checks/errors that can occur during view ops.  

Details about new errors added for 'strict_view' ops:
(1) a Flatten is only allowed if only the left-most input dim is sharded or all input dims are replicated
(2) when Flatten is applied on a sharded left-most input dim that is unevenly sharded, the Flatten is also illegal since it would likely change the unevenness and require resharding.


Follow ups
* we should decide what other view-like ops behavior should be (this PR only addressed the 'view' op

Fixes #143372

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

add 'strict_view' kwarg to the helper methods that implement
view/reshape op shard prop rules, so it can be decided op-by-op whether
to raise these new errors
enabled errors just for the 'view' op in this PR
added two specific checks/errors that can occur during view ops.

Details:

- View ops are never allowed to flatten a dimension that is unevenly
  sharded, since that would likely change the size/content of the
  local_tensor and require redistribute
- View ops are also never allowed to flatten two dims if the rightmost
  dim is a Shard() placment, becuase it would cause contiguity errors
  without redistribution

Notes:

- Disables support for `cartesian_prod` in dtensor ops test, as this op
  decomposes to an illegal view that only works by performing a
  redistribution.

Follow Ups:
- triage other view-like ops (besides aten::view) for using strict_view
- look for other gaps where view-like ops could still perform
  redistribution (ban them all, and document this)

Fixes #143372

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM! I think this PR captures some corner cases in view ops.

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

add 'strict_view' kwarg to the helper methods that implement
view/reshape op shard prop rules, so it can be decided op-by-op whether
to raise these new errors
enabled errors just for the 'view' op in this PR
added two specific checks/errors that can occur during view ops.

Details:

- View ops are never allowed to flatten a dimension that is unevenly
  sharded, since that would likely change the size/content of the
  local_tensor and require redistribute
- View ops are also never allowed to flatten two dims if the rightmost
  dim is a Shard() placment, becuase it would cause contiguity errors
  without redistribution

Notes:

- Disables support for `cartesian_prod` in dtensor ops test, as this op
  decomposes to an illegal view that only works by performing a
  redistribution.

Follow Ups:
- triage other view-like ops (besides aten::view) for using strict_view
- look for other gaps where view-like ops could still perform
  redistribution (ban them all, and document this)

Fixes #143372

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

add 'strict_view' kwarg to the helper methods that implement
view/reshape op shard prop rules, so it can be decided op-by-op whether
to raise these new errors
enabled errors just for the 'view' op in this PR
added two specific checks/errors that can occur during view ops.

Details:

- View ops are never allowed to flatten a dimension that is unevenly
  sharded, since that would likely change the size/content of the
  local_tensor and require redistribute
- View ops are also never allowed to flatten two dims if the rightmost
  dim is a Shard() placment, becuase it would cause contiguity errors
  without redistribution

Notes:

- Disables support for `cartesian_prod` in dtensor ops test, as this op
  decomposes to an illegal view that only works by performing a
  redistribution.

Follow Ups:
- triage other view-like ops (besides aten::view) for using strict_view
- look for other gaps where view-like ops could still perform
  redistribution (ban them all, and document this)

Fixes #143372

[ghstack-poisoned]

wanchaol commented:
> my first question is, are we sure its OK to regress this functionality (some users might be relying on view to implicitly redistribute)?
> 
> second question is, assuming we regress it, what's the right test coverage strategy (a) i thought about trying to modify test_dtensor_ops.py to let us specify manually which placements to test for a given op. I could also specify for failing placement what exception type we want. I briefly tried this and found its complicated to implement. (b) we can just xfail these ops in test_dtensor_ops and rely on specific tests in test_view_ops to cover view tests. Current preference is (b).

For 1. Yeah I think it's ok to regress because that's the intention of this PR. I'm not sure we should make it a BC breaking change or not, I feel not because this is more like a bug fix

For 2. I think (b) is fine. Ideally we can refactor test_dtensor_ops.py to not only check all sharding combinations of an op, but also have sth like a map to register the expected failure sharding combination, so that the rest of the sharding combinations can still be tested.

wconstab commented:
> For 2. I think (b) is fine. Ideally we can refactor test_dtensor_ops.py to not only check all sharding combinations of an op, but also have sth like a map to register the expected failure sharding combination, so that the rest of the sharding combinations can still be tested.

ok sgtm. I also liked this idea at first. I felt its a bit tricky to implement though.

I started to try hardcoding a list of supported shardings for a given op.  like [Replicate(), Replicate()], [Replicate(), Shard(0)] ...

But thats tricky too becuase different shapes of tensors are passed in and we have to generate sharding specs for those, so I can't just hardcode a list that bakes in Shard(2) when the input tensor might be 1-d or 2-d or 3-d.

wconstab committed:
Update on "[DTensor] Error on illegal view op during sharding prop"


Adds explicit error checking during sharding propagation for view ops
rather than relying on runtime errors during local op execution.

Before:
An error is thrown by aten.view op called by DTensor dispatch, because
the local shard size is incompatible with the (incorrectly calculated)
args to the view op.

`RuntimeError: shape '[384]' is invalid for input of size 512`

After:
We raise more specific errors for cases of incompatible view operations
during sharding propagation, before getting to runtime dispatch.

`RuntimeError: Attempted to flatten an unevenly sharded dimension, which would require resharding the input. Please explicitly redistribute the tensor instead.`

Change Summary:

add 'strict_view' kwarg to the helper methods that implement
view/reshape op shard prop rules, so it can be decided op-by-op whether
to raise these new errors
enabled errors just for the 'view' op in this PR
added two specific checks/errors that can occur during view ops.

Details:

- View ops are never allowed to flatten a dimension that is unevenly
  sharded, since that would likely change the size/content of the
  local_tensor and require redistribute
- View ops are also never allowed to flatten two dims if the rightmost
  dim is a Shard() placment, becuase it would cause contiguity errors
  without redistribution

Notes:

- Disables support for several ops in test_dtensor_ops.py test, which
  decompose to an illegal view that only works by performing a
  redistribution: cartesian_prod, flatten, ravel, reshape, reshape_as, view, view_as, take_along_dim, kron

Follow Ups:
- triage other view-like ops (besides aten::view) for using strict_view
- look for other gaps where view-like ops could still perform
  redistribution (ban them all, and document this)

Fixes #143372

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -f "infra issue? regular merge timed out"
